service/src/main/java/org/ehrbase/service/DirectoryServiceImp.java

```diff
@@ -163,16 +166,35 @@ public Folder update(UUID ehrId, Folder folder, ObjectVersionId ifMatches) {
 
     @Override
     public Folder update(UUID ehrId, Folder folder, ObjectVersionId ifMatches, UUID contributionId, UUID auditId) {
+
+        UUID uuid = UUID.fromString(ifMatches.getObjectId().getValue());
+
+        // Ensure UIDs matching
+        String folderUidValue = Optional.ofNullable(folder.getUid())
+                .map(UIDBasedId::getRoot)
+                .map(UID::getValue)
+                .orElse(null);
+
+        if (folderUidValue != null && !Objects.equals(uuid.toString(), folderUidValue)) {
+            throw new PreconditionFailedException(
+                    String.format("FOLDER uid %s does not match %s", folderUidValue, uuid));
+        }
+
         // validation
         ehrService.checkEhrExistsAndIsModifiable(ehrId);
-        if (!ehrFolderRepository.hasFolder(ehrId, EHR_DIRECTORY_FOLDER_IDX)) {
+
+        if (!ehrFolderRepository.hasFolderInEhrForVoId(ehrId, uuid)) {
```

Comment:
The check must me constrained to EHR_DIRECTORY_FOLDER_IDX

Reply:
 I've added the idx=1 constraint, but I'm not sue if this is needed at all.  If we check for ehr_id + vo_id this combination shall be unique regardless of the folder index 🤔.

Reply:
As of now there is just one vo_id per ehr.directory. The data model, however,  is already prepared for EHR.folders, so a vo_id is bound to a (ehr_id, folder_idx) pair.